pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -443,35 +451,97 @@ func convertRbacRulesToFilterConfig(
+	tcpConfig := listener.Filter{
+		Name: rbacTCPFilterName,
+		Config: util.MessageToStruct(&network_config.RBAC{
+			Rules:       config.Rules,
+			ShadowRules: config.ShadowRules,
+			StatPrefix:  rbacTCPFilterStatPrefix,
+		}),
+	}
+	rbacLog.Debugf("generated tcp filter config: %v", tcpConfig)
+	return &tcpConfig
+}
+
+// buildHTTPFilter builds the RBAC http filter that enforces the access control to the specified
+// service which is co-located with the sidecar proxy.
+func buildHTTPFilter(service *serviceMetadata, store model.IstioConfigStore) *http_conn.HttpFilter {
+	roles, bindings, err := rbacPolicyForService(service, store)
+	if err != nil {
+		rbacLog.Errorf("failed to get rbac policy: %v", err)
+		return nil
+	}
+
+	config := convertRbacRulesToFilterConfig(service, rbacOption{roles: roles, bindings: bindings, forTCPFilter: false})
+	rbacLog.Debugf("generated http filter config: %v", *config)
+	return &http_conn.HttpFilter{
+		Name:   rbacHTTPFilterName,
+		Config: util.MessageToStruct(config),
+	}
+}
 
+// convertRbacRulesToFilterConfig converts the current RBAC rules (ServiceRole and ServiceRoleBindings)
+// in service mesh to the corresponding proxy config for the specified service. The generated proxy config
+// will be consumed by envoy RBAC filter to enforce access control on the specified service.
+func convertRbacRulesToFilterConfig(service *serviceMetadata, option rbacOption) *http_config.RBAC {
 	rbac := &policyproto.RBAC{
 		Action:   policyproto.RBAC_ALLOW,
 		Policies: map[string]*policyproto.Policy{},
 	}
-
 	permissiveRbac := &policyproto.RBAC{
 		Action:   policyproto.RBAC_ALLOW,
 		Policies: map[string]*policyproto.Policy{},
 	}
 
-	for _, role := range roles {
-		enforcedPrincipals, permissivePrincipals := convertToPrincipals(roleToBinding[role.Name])
-		if len(enforcedPrincipals) == 0 && len(permissivePrincipals) == 0 {
-			rbacLog.Debugf("role skipped for no principals found")
-			continue
-		}
-
+	roleToBindings := roleToBindings(option.bindings)
+	for _, role := range option.roles {
 		rbacLog.Debugf("checking role %v", role.Name)
 		permissions := make([]*policyproto.Permission, 0)
 		for i, rule := range role.Spec.(*rbacproto.ServiceRole).Rules {
 			if service.match(rule) {
-				// Generate the policy if the service is matched to the services specified in ServiceRole.
-				rbacLog.Debugf("matched AccessRule[%d]", i)
-				permissions = append(permissions, convertToPermission(rule))
+				rbacLog.Debugf("rules[%d] matched", i)
+				if option.forTCPFilter {
+					if err := validateRuleForTCPFilter(rule); err != nil {
+						// It's a user misconfiguration if a HTTP rule is specified to a TCP service.
+						// For safety consideration, we ignore the whole rule which means no access is opened to
+						// the TCP service in this case.
+						rbacLog.Errorf("rules[%d] ignored, found HTTP only rule for a TCP service: %v", i, err)
```

Comment:
would be nice to have metrics at some point. Sicne you have pushContext variable here, you can add metrics here as well.

Secondly, even the minor inadvertent application of a http rbac to tcp will cause a lot of spam in the logs because of this. I suggest converting this error message to debug

Reply:
Thanks for the info, will add metrics later.